accepted/2021/default_string_globalization.md

```diff
@@ -0,0 +1,390 @@
+
+Obsoleting existing common APIs may also invalidate existing C# tutorials and introductory courses. For example, in Microsoft's own C# "Hello world" tutorial, [Step 4](https://docs.microsoft.com/dotnet/csharp/tour-of-csharp/tutorials/hello-world?tutorial-step=4) and [Step 6](https://docs.microsoft.com/dotnet/csharp/tour-of-csharp/tutorials/hello-world?tutorial-step=6) both contain calls to APIs this proposal intends to obsolete. It would be a poor experience for a novice developer to encounter a warning so early into their .NET journey. (This code sample focuses on Microsoft's own tutorial pages, but we should be mindful that these types of samples are probably very widespread.)
+
+```cs
+/* Sample code from Step 4, containing calls to proposed obsolete methods */
+
+Console.WriteLine(sayHello.ToUpper());
+Console.WriteLine(sayHello.ToLower());
+
+/* Sample code from Step 6, containing calls to proposed obsolete methods */
+
+string songLyrics = "You say goodbye, and I say hello";
+Console.WriteLine(songLyrics.StartsWith("You"));
+Console.WriteLine(songLyrics.StartsWith("goodbye"));
+
+Console.WriteLine(songLyrics.EndsWith("hello"));
+Console.WriteLine(songLyrics.EndsWith("goodbye"));
+```
+
+As a minor concern, developers working in an interactive debugger (think VS's Immediate window) will not see any obsoletion messages that we put on our public API surface. If they enter `theStr.IndexOf("foo")` into the Immediate window, that will evaluate in a linguistic, current-culture context. This is not strictly a behavioral change from previous runtimes, but it could be slightly confusing to developers who have the old patterns engrained into their muscle memory and who rely on our warnings and fixers for guidance.
+
+Finally, we cannot rely on analyzers and fixers alone to detect when a linguistic, current-culture call is about to be made. If a `string` instance is passed around as `object` and some consumer has code which reads `if (obj is IComparable comparable) { comparable.CompareTo(...); }`, there's not much we can do about this. If it's too unpalatable to change the underlying behavior of `string` itself, then the best we can do is direct people to use a modern API surface and generic collections, avoiding non-generic methods like `Array.Sort(Array)`.
+
+## Future work and ecosystem implications
+
+This work has implications for GUI stacks, such as WPF and WinForms. GUI scenarios often involve performing linguistic, current-culture collation of data, such as alphabetically sorting entries in a dropdown list. It should be expected that WinForms, WPF, and similar code bases will be disproportionately hit with warnings. They have a few options available to them. In descending order of preference:
```

Comment:
> similar code bases will be disproportionately hit with warnings

Not only hit with warnings, but there's a good chance pushing them to use ordinal behavior will actually introduce bugs.

Reply:
That's the weird part about all of this. The rule of thumb is: if you're sorting data for display directly in client UI, the default (_current culture_) behavior is probably appropriate. If you're on a server, a current culture comparison is probably not appropriate. So it largely depends on the context.

However, there is one universal truth: __the pattern `if (string.Compare(...) == 0)` is almost certainly a bug__, regardless of context. There are exceptions, sure, but they're very rare.

Maybe the answer is that we don't hide the original methods, but we instead query what you're doing with the result. If we see a comparison result `== 0` or `!= 0` we flag it. This will catch fewer misusages of the comparison APIs, but experience tells me that the warnings will be more actionable / of higher quality.